### ISSUE TITLE
Use mocking of HTTP responses for unit tests

### ISSUE THREAD

**Issue by nuest** [ropensci/suppdata#36]
See extensive instructions at https://ropensci.github.io/http-testing-book/

**Comment by willpearse** [comment]
Nice, this would make life easier. Thanks Daniel. Maybe something for the hack-a-thon :D

**Comment by willpearse** [comment]
Looking into this more, in the context of the R-devel thread, I'm not sure this would help us very much. It seems that we can't mock writing to disk yet (https://github.com/ropensci/webmockr/issues/57) - but I could be wrong. What do you think?

**Comment by willpearse** [comment]
OK, honestly, at this point I feel like I'm spending way longer thinking about this problem than is helpful, and it's slowing down development. In the last commit (https://github.com/ropensci/suppdata/commit/26a6e1388f96fcbfa56ad756861ae7c2d2bdcd21), I added a randomisation thing to the tests so only 1/5 times is a set of tests for a particular publisher run on CRAN. This means we're still running some of the tests, but a random subset, and as such even when loads of tests are run in parallel on their server we're unlikely to get this problem. I'm going to send this to them and see what CRAN think. I have reached out to others who I thought might have faced this problem, and their solutions in the past have been to skip all the tests, so I feel we're doing OK here.

**Comment by nuest** [comment]
We can't mock writing to disc. But we could add a `test.zip`, with a few of minimal files. The binary in the repo is of course not perfect, but it should be quite small. This could allow us to have tests for:

- extraction features and accessing specific files from a ZIP file via `si` option
- file listing via `list` option
- `save.name` option
- `dir` option

**Comment by willpearse** [comment]
I don't have a problem with doing this, but I don't see it as a high priority. I would prefer we increased test coverage (89% is good, but 100% is better) rather than wrote additional tests for the offline components of the package. Right now we have CI tests on Travis, OpenCPU, and AppVeyor. I'm not concerned that the tests aren't running on CRAN because CRAN aren't concerned, if that makes sense?

**Comment by willpearse** [comment]
Something to report here... I've used `vcr` to do this for another package (https://github.com/pearselab/SymbiotaR2/blob/master/tests/test-all.R) but it doesn't seem to work well if the `GET` statement downloads a file. The cache step caches the call, but not the file, and thus I could only get it work if I kept the output from the call all within R's memory. That makes sense, but would make it impossible for us to use `vcr` (right now) and keep `suppdata`'s caching feature. I don't want to drop a very useful feature to get the tests to work, so I'm closing this issue for now (with hopes that, one day, we'll be able to do this).
